### ISSUE TITLE
Wrong file location for ".next" dir when used in monorepo

### ISSUE THREAD

**Issue by ibraelillo** [serverless-stack/open-next#7]
Hi guys, I've been actively using open-next in the last few days and I see the following problem: 

When the next app is inside the packages folder in a monorepo structure, the ".next" folder is not correctly resolved when outputFileTracing is set to the monorepo's root:

![Captura desde 2022-12-14 11-18-45](https://user-images.githubusercontent.com/3011753/207569097-a93a927e-433f-4348-b347-01986939692b.png)

The output file structure in the left , and the code loading ".next" on the right, we see clearly that it won't work because the .next folder is inside a different file structure. 

![Captura desde 2022-12-14 11-17-24](https://user-images.githubusercontent.com/3011753/207568859-2419aadc-05f2-4eaf-bc49-c051509b9bbc.png)

Maybe doing something like:

```typescript
// build.ts
// before  nextBuild

const { experimental } = require('./next.config.js')

const monorepoRootPrefix = experimental.outputFileTracingRoot ? path.relative(experimental.outputFileTracingRoot, appPath) : __dirname)

// then, on the server-function build:

const result = buildSync({
    entryPoints: [
      path.resolve(__dirname, "../assets/server-adapter.cjs")
    ],
    bundle: true,
    target: "node16",
    platform: "node",
    external: ["next", "aws-sdk"],
    outfile: path.join(outputPath, "index.cjs"),
    format: "cjs",
    define: {
       monorepoRootPrefix: JSON.stringify(monorepoRootPrefix)
    }
  });

// server-adapter.cjs
const nextDir = path.join(monorepoRootPrefix, ".next");

```
This way we could not only resolve monorepo root , but also hoisted node_modules at the level of the monorepo  root . 

In any case, when the app is far down inside a monorepo structure, the resolution of the next folder is not working well inside the lambda.

**Comment by khuezy** [comment]
Did you get your example setup to work? `@vercel/next` build doesn't work well w/ symlinks.

**Comment by ibraelillo** [comment]
I've copied the build.ts locally and did what I've posted before, and it worked!
![Captura desde 2022-12-15 14-35-34](https://user-images.githubusercontent.com/3011753/207872546-5ea67ded-c766-40f6-b3c2-c3c16e2b7a48.png)

to remove the server file:
![Captura desde 2022-12-15 14-36-16](https://user-images.githubusercontent.com/3011753/207872694-6fc9c7f6-6753-41aa-ad3c-9db6a1ebdddd.png)

to build the layer for the lambda to reduce size:
![Captura desde 2022-12-15 14-40-08](https://user-images.githubusercontent.com/3011753/207873523-5fe16675-ffb5-42b3-956d-4f2bfd8eaae0.png)

and the server-handler:
![Captura desde 2022-12-15 14-41-03](https://user-images.githubusercontent.com/3011753/207873670-c9f0e989-810f-42c2-8c14-e962d4254aa7.png)
![Captura desde 2022-12-15 14-41-54](https://user-images.githubusercontent.com/3011753/207873853-e104a472-605d-4b7e-8661-990be63afa8c.png)

the full code for the build.mjs is (I've took the built code for facility reasons):
```javascript
import fs from 'fs'
import path from 'path'
import {buildSync} from 'esbuild'
import url from "url";
// @ts-ignore @vercel/next does not provide types
//import { build as nextBuild } from '@vercel/next'

const appPath = process.cwd()
const outputDir = '.open-next'
let monorepoRootPrefix = '.'


const __dirname = url.fileURLToPath(new URL('.', import.meta.url))

export async function build() {
    // Check running inside Next.js app
    if (!fs.existsSync(path.join(appPath, 'next.config.js'))) {
        console.error(
            'Error: next.config.js not found. Please make sure you are running this command inside a Next.js app.'
        )

        process.exit(1)
    }

    const {experimental} = await import(`${appPath}/next.config.js`).then(d => d.default)

    if (experimental.outputFileTracingRoot) {

        const root = path.isAbsolute(experimental.outputFileTracingRoot) ? experimental.outputFileTracingRoot : path.resolve(appPath, experimental.outputFileTracingRoot)

        monorepoRootPrefix = path.relative(root, appPath)
    }

    const nextBuild = await import('@vercel/next').then((d) => d.build)

    // Build app
    const ret = await nextBuild({
        files: [],
        workPath: appPath,
        entrypoint: 'next.config.js',
        config: {},
        meta: {
            skipDownload: true
        }
    })

    // Prepare output directory
    fs.rmSync(outputDir, {recursive: true, force: true})

    // Create server Lambda function bundle
    createServerBundle()

    createLayer()

    createMiddlewareBundle(ret.output.middleware.files['index.js'].data)

    createAssets()
}

function createServerBundle() {
    console.debug('Bundling server function...')

    // Create output folder
    const outputPath = path.join(outputDir, 'server-function')

    fs.mkdirSync(outputPath, {recursive: true})

    fs.cpSync(path.join(appPath, '.next/standalone'), path.join(outputPath), {
        recursive: true
    })

    // Create the Lambda handler inside the .next/standalone directory
    // Note: .next/standalone already has a Node server "server.js",
    // replace it with Lambda handler
    fs.rmSync(path.join(outputPath, monorepoRootPrefix, 'server.js'), {force: true})

    const result = buildSync({
        entryPoints: [path.resolve(__dirname, './assets/server-handler.js')],
        bundle: true,
        target: 'node16',
        platform: 'node',
        external: ['next', 'aws-sdk', 'sharp'],
        preserveSymlinks: false,
        outfile: path.join(outputPath, 'index.js'),
        format: 'cjs',
        define: {
            APP_ROOT: JSON.stringify(monorepoRootPrefix)
        }
    })

    if (result.errors.length > 0) {
        result.errors.forEach((error) => console.error(error))

        throw new Error('There was a problem bundling the server handler.')
    }
}

function createMiddlewareBundle(src) {
    console.debug('Bundling middleware edge function...')

    // Create output folder
    const outputPath = path.join(outputDir, 'middleware-function')

    fs.mkdirSync(outputPath, {recursive: true})

    // Create the middleware code
    const buildTempPath = path.join(outputDir, '.middleware-function')

    fs.mkdirSync(buildTempPath, {recursive: true})

    fs.writeFileSync(path.join(buildTempPath, 'middleware.js'), src)

    fs.copyFileSync(
        path.join(__dirname, './assets/middleware-handler.js'),
        path.join(buildTempPath, 'middleware-handler.js')
    )

    // Create a directory that we will use to create the bundled version
    // of the "core server build" along with our custom Lamba server handler.
    const result = buildSync({
        entryPoints: [path.join(buildTempPath, 'middleware-handler.js')],
        target: 'esnext',
        format: 'esm',
        platform: 'node',
        metafile: true,
        bundle: true,
        write: true,
        allowOverwrite: true,
        outfile: path.join(outputPath, 'index.js')
    })

    if (result.errors.length > 0) {
        result.errors.forEach((error) => console.error(error))

        throw new Error('There was a problem bundling the middleware handler.')
    }

    // Create package.json
    fs.writeFileSync(path.join(outputPath, 'package.json'), '{"type":"module"}')
}

function createAssets() {
    console.debug('Bundling assets...')

    // Create output folder
    const outputPath = path.join(outputDir, 'assets')

    fs.mkdirSync(outputPath, {recursive: true})

    // Next.js outputs assets into multiple files. Copy into the same directory.
    // Copy over:
    // - .next/BUILD_ID => _next/BUILD_ID
    // - .next/static   => _next/static
    // - public/*       => _next/*
    fs.copyFileSync(
        path.join(appPath, '.next/BUILD_ID'),
        path.join(outputPath, 'BUILD_ID')
    )

    fs.cpSync(
        path.join(appPath, '.next/static'),
        path.join(outputPath, '_next', 'static'),
        {recursive: true}
    )

    fs.cpSync(path.join(appPath, 'public'), path.join(outputPath, '_next'), {
        recursive: true
    })
}

function createLayer() {

    if (monorepoRootPrefix !== '.') {
        console.debug('Bundling layer...')

        // Create output folder
        const outputPath = path.join(outputDir, 'layer', 'nodejs')

        fs.mkdirSync(outputPath, {recursive: true})

        fs.cpSync(path.join(appPath, '.next/standalone/node_modules'), path.join(outputPath), {
            recursive: true
        })

        // possibly you will have a node_modules at the level of .next, so try copying it to the root because lambda environment
        // does not support yet pnp modules.
        // this copy will remove naturally the symlink created by yarn/pnpm
        fs.cpSync(path.join(appPath, '.next/standalone', monorepoRootPrefix, 'node_modules'), outputPath, {
            recursive: true
        })
    }
}

build()
```

**Comment by khuezy** [comment]
Nicely done! Do you mind putting up a PR for @fwang to review? I can test it on end too.

**Comment by fwang** [comment]
@ibraelillo Appreciate the detailed write up. I implemented ur solution with 2 minor tweaks:

1. Auto-detecting the monorepo root, so as a user, u don't need to add `outputFileTracingRoot` to your next config. `outputFileTracingRoot` is not required when deploying to Vercel. So I'm assuming most users don't have that already specified in their config. This allow them to deploy their Next.js as is.
2. Instead of outputting `index.mjs` to `.open-next/server-function/packages/xxxx`, I ended up moving `.open-next/server-function/packages/xxxx` to `.open-next/server-function`.
